bip-0008.mediawiki

```diff
@@ -169,6 +164,29 @@ And ACTIVE and FAILED are terminal states, which a deployment stays in once they
         }
     }
 
+Note that blocks received while in MUST_SIGNAL must be checked to ensure that the threshold will be reached by the end of the period. For example:
+
+    if (GetStateForBlock(block) == MUST_SIGNAL) {
+        if (block.nVersion & 0xE0000000 != 0x2000000 || block.nVersion >> bit) & 1 != 1) {
+            nonsignal = 1;
+            count = block.nHeight % 2016;
+            walk = block;
+            while (count > 0 && nonsignal + count + threshold > 2016) {
+                --count;
+                walk = walk.parent;
+                if (walk.nVersion & 0xE0000000 != 0x2000000 || walk.nVersion >> bit) & 1 != 1) {
+                    ++nonsignal;
+                    if (nonsignal + threshold > 2016) break;
+                }
+            }
+            if (nonsignal + threshold > 2016) {
+                return state.Invalid(BlockValidationResult::RECENT_CONSENSUS_CHANGE, "blk-bip8-must-signal");
+            }
+        }
+    }
```

Comment:
0x2... misses a 0 and the parantheses are wrong (there's at least one opening paranthesis missing). I tested this code a little bit and it seems correct. Nit: Could perhaps be a bit simplified by moving the break condition to the while conditions and moving the `nonsignal + count + threshold > 2016` condition somewhere else because the LHS is constant across loop iterations. Perhaps something like

```
walk = block;
if (count + threshold >= 2016) {
    while (count > 0  && nonsignal + threshold <= 2016)  {
        ...
        if ... {
            ++nonsignal;
        }
    }
    if (nonsignal + threshold > 2016) {
        return Invalid;
    }
}
```

Generally the suggested changes look very reasonable but I can't speak to the p2p relay issues.

Reply:
`count` decreases every iteration but `nonsignal` only increases sometimes, so the LHS is decreasing, which lets the loop exit early (if you've only seen 5 nonsignalling blocks, and there's only 95 more blocks to look at, you'll never get 101 nonsignalling blocks so no need to look any further). Not really very useful to include in the spec though. No real need to skip the loop if the current block signals either, I think which might make the spec a bit simpler.

Reply:
Updated this

Reply:
Hmm, not sure why I said the LHS would be constant. But what I meant is that you can leave count out of this as I did in my pseudocode.

Anyway, the new code passes my tests. So ACK that.

> No real need to skip the loop if the current block signals either,

Fwiw, the updated code still seems to that with the `else if` block.